orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/PatchManifestContext.java

```diff
@@ -43,10 +44,51 @@
 
   @Builder.Default private boolean skipExpressionEvaluation = false;
 
+  @Builder.Default
+  private PatchManifestContext.Options options = PatchManifestContext.Options.builder().build();
+
+  @Builder(builderClassName = "OptionsBuilder", toBuilder = true)
+  @JsonDeserialize(builder = PatchManifestContext.Options.OptionsBuilder.class)
+  @Value
+  static class Options {
+    @Builder.Default private MergeStrategy mergeStrategy = MergeStrategy.STRATEGIC;
+    @Builder.Default private boolean record = true;
+
+    @JsonPOJOBuilder(withPrefix = "")
+    static class OptionsBuilder {}
+  }
+
+  public enum MergeStrategy {
+    @JsonProperty("strategic")
+    STRATEGIC,
+
+    @JsonProperty("json")
+    JSON,
+
+    @JsonProperty("merge")
+    MERGE
+  }
+
   @Nullable
   @Override
   public List<Map<Object, Object>> getManifests() {
-    return Collections.singletonList(patchBody);
+    /*
+     * Clouddriver expects patchBody to be a List when the mergeStrategy is json, and a Map otherwise.
+     * Prior to Spinnaker 1.15, Deck sent either a List or Map depending on merge strategy, and Orca
+     * deserialized it as an Object. In Spinnaker 1.15 and 1.16, a bug was introduced where patchBody
+     * was deserialized as a Map regardless of merge strategy. Starting in 1.17, Deck will send a List
+     * regardless of merge strategy, but we should handle receiving a Map from pipelines configured
+     * in 1.15 and 1.16.
+     */
+    if (patchBody == null) {
+      return null;
+    }
+    if (patchBody instanceof List) {
+      return (List) patchBody;
+    }
+    List<Map<Object, Object>> manifests = new ArrayList<>();
```

Comment:
This could be simplified to `ImmutableList.of((Map<Object, Object>) patchBody)`; this would of course also make the return value immutable, which might be a good side-effect.

(I tend to avoid having a function sometimes return a mutable list and sometimes an immutable one, so if you take this suggestion it might also be good to change the above return to `ImmutableList.copyOf((List) patchBody)`).

And similarly in the null case, we might want to return `ImmutableList.of()` and annotate the function `@Nonnull` to make it easier on callers, as I don't think there would ever be a real difference in meaning when this returns `null` vs. an empty list.

Reply:
Loooove it, thank you, updated on all counts!